ADAL/src/ADAuthenticationParameters+Internal.m

```diff
@@ -95,79 +96,175 @@ + (ADAuthenticationError *)invalidHeader:(NSString *)headerContents
-                }
-            }
+            break;
+        }
+        
+        NSMutableDictionary *nextParameters = [self extractParameters:items[i]];
+        [parameters addEntriesFromDictionary:nextParameters];
+    }
+    
+    return parameters;
+}
+
+#pragma mark - Private
+
++ (NSMutableArray *)extractItems:(NSString *)string
+{
+    NSMutableArray *items = [NSMutableArray new];
+    
+    NSInteger rightIndex = 0;
+    NSInteger leftIndex = 0;
+    while (rightIndex < string.length)
+    {
+        unichar c = [string characterAtIndex:rightIndex];
+        
+        // Start of a quaoted string, lets get last index of it and continue iteration.
+        if (c == '"' || c == '\'')
+        {
+            rightIndex = [self quotedStringLastIndex:string startIndex:rightIndex];
+            rightIndex++;
+            continue;
+        }
+        
+        if (c == ',')
+        {
+            NSUInteger len = rightIndex - leftIndex;
+            NSRange range = NSMakeRange(leftIndex, len);
+            NSString *item = [string substringWithRange:range];
+            
+            [items addObject:item];
+            
+            leftIndex = rightIndex + 1;
+        }
+        
+        rightIndex++;
+    }
+    
+    if (leftIndex < rightIndex)
+    {
+        NSUInteger len = rightIndex - leftIndex;
+        NSRange range = NSMakeRange(leftIndex, len);
+        NSString *item = [string substringWithRange:range];
+        
+        [items addObject:item];
+    }
+    
+    // Check for invalid parameters/state.
+    
+    if (leftIndex == rightIndex)
+    {
+        // Comma was not followed by a text -- invalid header.
+        return nil;
+    }
+    
+    for (NSString *item in items) {
+        if ([NSString adIsStringNilOrBlank:item]) {
+            // Blank item found -- invalid header.
+            return nil;
         }
     }
     
-    if (rgError)
+    return items;
+}
+
++ (NSUInteger)quotedStringLastIndex:(NSString *)headerContents startIndex:(NSUInteger)startIndex
+{
+    NSUInteger lastIndex = startIndex;
+    
+    unichar quotedChar = [headerContents characterAtIndex:startIndex];
+    if (quotedChar == '"' || quotedChar == '\'')
     {
-        //The method below will log internally the error:
-        adError =[ADAuthenticationError errorFromNSError:rgError errorDetails:rgError.description correlationId:nil];
+        NSUInteger nextIndex = startIndex + 1;
+        while (nextIndex < headerContents.length) {
+            unichar c = [headerContents characterAtIndex:nextIndex];
+            
+            if (c == quotedChar) {
+                lastIndex = nextIndex;
+                break;
+            }
+            
+            nextIndex++;
+        }
     }
     
-    if (error)
+    return lastIndex;
+}
+
++ (BOOL)isParameter:(NSString *)string
+{
+    NSRange range = [string rangeOfString:ExtractionExpression options:NSRegularExpressionSearch];
+    BOOL result = range.location == 0;
```

Comment:
Any reason to not just return here?

Reply:
I used it for debugging and forgot to remove. Will replace it with return.